src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+
+/**
+ * Checks if arguments of methods are properly configured.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class CheckArgumentsValidityPass extends AbstractRecursivePass
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        $arguments = $value->getArguments();
+        if ($arguments !== array_values($arguments)) {
+            throw new RuntimeException(sprintf('Arguments\'s keys of the "%s" service must be integers or names of constructor\'s parameters prefixed by the "$" symbol. Check your service definition.', $this->currentId));
```

Comment:
Ok, a few things about this :)

**1**

Take this setup (this is my setup from part A here https://github.com/symfony/symfony/pull/21383#issuecomment-275737454)

```php
class FunGenerator
{
    public function __construct($arg1, $arg2)
    {
    }
}
```

```yml
services:
    fun_generator:
        class: AppBundle\Services\FunGenerator
        arguments:
            $arg2: 'foo_bar'
```

In this case, you WILL hit the above exception. Yay! That's great, because I think this code was added to give the user a better exception with this mistake (which is what I was hoping for!). But, the exception is quote confusing:

> Arguments's keys of the "fun_generator" service must be integers or names of constructor's parameters prefixed by the "$" symbol. Check your service definition.

What happened is, `ResolveNamedArgumentsPass` correctly maps `$arg2` as the `[1]` index... but index `[0]` is missing. So, it hits this exception... but the message is obviously not quite right. This relates to my next points.

**2**
Can we actually get to this point in the code with a key that is still prefixed with $? I think the `ResolveNamedArgumentPass` will already have converted all `$` args to an index arg (or thrown an exception if there is no matching arg name).

**3**

Also, can you even get here by using a non-integer key (which is now deprecated)? The `ResolveNamedArgumentPass` (right after the deprecation trigger) is already converting all non-integer, non-$-prefix arguments to their index/integer value. So, by the time we get here, I don't think it's possible to have *anything* except for integer keys. We're already "correcting" them earlier, which means that I don't think this block is doing anything.


tl;dr; Exactly what situations is this pass meant to catch? Let's make sure our message explains the correct situation.